go-controller/pkg/ovn/pods_retry.go

```diff
@@ -0,0 +1,205 @@
+				}
+			}
+		} else {
+			klog.V(5).Infof("%s retry pod not after timer yet, time: %s", podDesc, podTimer)
+		}
+	}
+}
+
+// checkAndDeleteRetryPod deletes a specific entry from the map, if it existed, returns true
+func (oc *Controller) checkAndDeleteRetryPod(pod *kapi.Pod) bool {
+	oc.retryPodsLock.Lock()
+	defer oc.retryPodsLock.Unlock()
+	key := getPodNamespacedName(pod)
+	if _, ok := oc.retryPods[key]; ok {
+		delete(oc.retryPods, key)
+		return true
+	}
+	return false
+}
+
+// checkAndSkipRetryPod sets a specific entry from the map to be ignored for subsequent retries
+// if it existed, returns true
+func (oc *Controller) checkAndSkipRetryPod(pod *kapi.Pod) bool {
+	oc.retryPodsLock.Lock()
+	defer oc.retryPodsLock.Unlock()
+	key := getPodNamespacedName(pod)
+	if entry, ok := oc.retryPods[key]; ok {
+		entry.ignore = true
+		return true
+	}
+	return false
+}
+
+// unSkipRetryPod ensures a pod is no longer ignored for retry loop
+func (oc *Controller) unSkipRetryPod(pod *kapi.Pod) {
+	oc.retryPodsLock.Lock()
+	defer oc.retryPodsLock.Unlock()
+	key := getPodNamespacedName(pod)
+	if entry, ok := oc.retryPods[key]; ok {
+		entry.ignore = false
+	}
+}
+
+// initRetryAddPod tracks a pod that failed to be created to potentially retry later (needsAdd = true)
+// initially it is marked as skipped for retry loop (ignore = true)
+func (oc *Controller) initRetryAddPod(pod *kapi.Pod) {
+	oc.retryPodsLock.Lock()
+	defer oc.retryPodsLock.Unlock()
+	key := getPodNamespacedName(pod)
+	if entry, ok := oc.retryPods[key]; ok {
+		entry.timeStamp = time.Now()
+		entry.pod = pod
+		entry.needsAdd = true
+	} else {
+		oc.retryPods[key] = &retryEntry{pod, time.Now(), 1, true, true, nil}
+	}
+}
+
+// initRetryDelPod tracks a pod that failed to be deleted to potentially retry later (needsDel != nil)
+// initially it is marked as skipped for retry loop (ignore = true)
+func (oc *Controller) initRetryDelPod(pod *kapi.Pod) {
+	oc.retryPodsLock.Lock()
+	defer oc.retryPodsLock.Unlock()
+	key := getPodNamespacedName(pod)
+	if !util.PodWantsNetwork(pod) {
+		// create dummy logicalPortInfo for host-networked pods
+		mac, _ := net.ParseMAC("00:00:00:00:00:00")
+		oc.logicalPortCache.add("host-networked", getPodNamespacedName(pod),
+			"host-networked", mac, []*net.IPNet{})
+	}
+	portInfo, _ := oc.logicalPortCache.get(key)
+	if entry, ok := oc.retryPods[key]; ok {
+		entry.timeStamp = time.Now()
+		entry.needsDel = portInfo
+	} else {
+		oc.retryPods[key] = &retryEntry{pod, time.Now(), 1, true, false, portInfo}
+	}
+}
+
+func (oc *Controller) removeDeleteRetry(pod *kapi.Pod) {
+	oc.retryPodsLock.Lock()
+	defer oc.retryPodsLock.Unlock()
+	key := getPodNamespacedName(pod)
+	if entry, ok := oc.retryPods[key]; ok {
+		entry.needsDel = nil
+	}
+}
+
+func (oc *Controller) getPodRetryEntry(pod *kapi.Pod) *retryEntry {
+	oc.retryPodsLock.Lock()
+	defer oc.retryPodsLock.Unlock()
+	key := getPodNamespacedName(pod)
+	if entry, ok := oc.retryPods[key]; ok {
+		entryCopy := entry
```

Comment:
this isn't a copy right? It's just a new pointer pointing to the same entry.

Reply:
haha yeah I had the same question here: https://github.com/ovn-org/ovn-kubernetes/pull/1121#discussion_r813367665

It seems like golang usually does a shallow copy but in this case it would hold a shallow copy of the address to the retryEntry struct, I should probably do the same thing you did, copy actual value return address.

Reply:
done.